### ISSUE TITLE
`assistant_mask` only zeros

### ISSUE THREAD

**Issue by ValeKnappich** [OpenPipe/ART#127]
I am trying to train on my own task using ART but get this after the rollout:

> There are no assistant logprobs to train on. Did you forget to include at least one Choice in Trajectory.messages_and_choices?

I believe this is because the `assistant_mask` is not computed correctly; when I debug I see it contains only zeros. Looking at the tokenization code, I don't see how this is supposed to work for the regular messages format (`[{"role": "assistant", "content": "..."}, ...]`)

https://github.com/OpenPipe/ART/blob/main/src/art/local/tokenize.py#L136-L160

```py
    assistant_mask = [0] * len(token_ids)
    for message_or_choice in trajectory.messages_and_choices:
        if isinstance(message_or_choice, dict):
            continue
        choice = message_or_choice
        assert choice.logprobs, "Chat completion choices must have logprobs"
        token_logprobs = choice.logprobs.content or choice.logprobs.refusal or []
        sentinal_index = token_ids.index(sentinal_token_id)
        token_ids[sentinal_index : sentinal_index + 1] = (
            int(token_logprob.token.split(":")[1]) for token_logprob in token_logprobs
        )
        logprobs[sentinal_index : sentinal_index + 1] = (
            token_logprob.logprob for token_logprob in token_logprobs
        )
        assistant_mask[sentinal_index : sentinal_index + 1] = [1] * len(token_logprobs)
    return TokenizedResult(
        trajectory=trajectory,
        advantage=advantage,
        chat=chat,
        tokens=[tokenizer.decode(token_id) for token_id in token_ids],
        token_ids=token_ids,
        input_pos=list(range(len(token_ids))),
        assistant_mask=assistant_mask,
        logprobs=logprobs,
    )
```

Does the `art.Trajectory` require a different messages format? Why are logprobs mandatory? What is the `sentinal_token` supposed to be?

@bradhilton You seem to have written most of the tokenization code, perhaps you could shed some light on this?

**Comment by bradhilton** [comment]
@ValeKnappich Are you including a `Choice` object in `trajectory.messages_and_choices`? You can share your code for building a trajectory here to help me understand your issue.

**Comment by ValeKnappich** [comment]
@bradhilton Thanks for the quick reply. I do not include a choice object as of now. I simply set messages_and_choices in this format:

```py
traj = art.Trajectory(
    messages_and_choices=[
        {"role": "user", "content": ""},
        {"role": "assistant", "content": ""},
    ],
    ...
)
```

Upon reading further, it seems like assistant responses are only recognized as such if they are a `Choice` object. Furthermore, the approach seems to rely entirely on the logprobs, which is weird because they are not returned by default by vllm. It seems to work if we include the logprobs (either post-hoc as below or by requesting them from the API):

```py
import art
import art.local
import art.local.tokenize
from transformers import AutoTokenizer, PreTrainedTokenizerBase
from openai.types.chat.chat_completion import Choice, ChoiceLogprobs, ChatCompletionMessage
from openai.types.chat.chat_completion_token_logprob import ChatCompletionTokenLogprob

tokenizer: PreTrainedTokenizerBase = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-3B-Instruct")

messages_only_dicts = [
    {"role": "system", "content": "system message"},
    {"role": "user", "content": "user message"},
    {"role": "assistant", "content": "assistant message"},
    {"role": "user", "content": "user message"},
    {"role": "assistant", "content": "assistant message"},
]

messages_empty_logprobs = [
    {"role": "system", "content": "system message"},
    {"role": "user", "content": "user message"},
    Choice(
        message=ChatCompletionMessage(role="assistant", content="assistant message"),
        finish_reason="stop",
        index=0,
        logprobs=ChoiceLogprobs()
    ),
    {"role": "user", "content": "user message"},
    Choice(
        message=ChatCompletionMessage(role="assistant", content="assistant message"),
        finish_reason="stop",
        index=0,
        logprobs=ChoiceLogprobs()
    )
]

messages_dummy_logprobs = [
    {"role": "system", "content": "system message"},
    {"role": "user", "content": "user message"},
    Choice(
        message=ChatCompletionMessage(role="assistant", content="assistant message"),
        finish_reason="stop",
        index=0,
        logprobs=ChoiceLogprobs()
    ),
    {"role": "user", "content": "user message"},
    Choice(
        message=ChatCompletionMessage(role="assistant", content="assistant message"),
        finish_reason="stop",
        index=0,
        logprobs=ChoiceLogprobs()
    )
]
for msg in messages_dummy_logprobs:
    if not isinstance(msg, Choice):
        continue
    assert msg.logprobs
    msg.logprobs.content = [
        ChatCompletionTokenLogprob(token=f":{token_id}", logprob=0.0, top_logprobs=[])
        for token_id in tokenizer.encode(msg.message.content)
    ]


for messages, name in (
    (messages_only_dicts, "only_dicts"), 
    (messages_empty_logprobs, "empty logprobs"), 
    (messages_dummy_logprobs, "dummy logprobs")
):
    print("\n\n", name)

    traj = art.Trajectory(messages_and_choices=messages, reward=0.0)

    tokenized = art.local.tokenize.tokenize_trajectory(tokenizer, traj, 0.0)

    for mask, token_id in zip(tokenized.assistant_mask, tokenized.token_ids):
        print(mask, tokenizer.decode(token_id).replace("\n", "\\n"))
```


```
 only_dicts
0 <|im_start|>
0 system
0 \n
0 system
0  message
0 <|im_end|>
0 \n
0 <|im_start|>
0 user
0 \n
0 user
0  message
0 <|im_end|>
0 \n
0 <|im_start|>
0 assistant
0 \n
0 assistant
0  message
0 <|im_end|>
0 \n
0 <|im_start|>
0 user
0 \n
0 user
0  message
0 <|im_end|>
0 \n
0 <|im_start|>
0 assistant
0 \n
0 assistant
0  message
0 <|im_end|>
0 \n


 empty logprobs
0 <|im_start|>
0 system
0 \n
0 system
0  message
0 <|im_end|>
0 \n
0 <|im_start|>
0 user
0 \n
0 user
0  message
0 <|im_end|>
0 \n
0 <|im_start|>
0 assistant
0 \n
0 <|im_end|>
0 \n
0 <|im_start|>
0 user
0 \n
0 user
0  message
0 <|im_end|>
0 \n
0 <|im_start|>
0 assistant
0 \n
0 <|im_end|>
0 \n


 dummy logprobs
0 <|im_start|>
0 system
0 \n
0 system
0  message
0 <|im_end|>
0 \n
0 <|im_start|>
0 user
0 \n
0 user
0  message
0 <|im_end|>
0 \n
0 <|im_start|>
0 assistant
0 \n
1 assistant
1  message
0 <|im_end|>
0 \n
0 <|im_start|>
0 user
0 \n
0 user
0  message
0 <|im_end|>
0 \n
0 <|im_start|>
0 assistant
0 \n
1 assistant
1  message
0 <|im_end|>
0 \n
```

**Comment by bradhilton** [comment]
Yes, you need the choice object with logprobs. The vLLM server ART serves for you should always return logprobs by default.

**Comment by ValeKnappich** [comment]
I see. Is there a particular reason to enforce logprobs here? Is it just an easy way to create the assistant_mask? AFAIU the vllm logprob values are not used otherwise

**Comment by bradhilton** [comment]
@ValeKnappich the logprobs are used during the training step with PPO style clipping to ensure the policy doesn't stray too far from the existing policy. I understand that passing a Choice object may be inconvenient is some cases, and I may work on an alternative that would allow just passing standard assistant message dictionaries. Is this really important for what you're trying to do or just something you're curious to understand?

**Comment by ValeKnappich** [comment]
My training seems to run fine with the Choice objects, I was just trying to understand if the logprobs are only a convenient way to compute the assistant mask, or if they are actually used later on. Reusing them to compute the KL loss makes sense, but the current implementation seems to call the model again for that (see [here](https://github.com/OpenPipe/ART/blob/f36cbcb97b278dcbda753481f407b8ea081c5f42/src/art/local/train.py#L97)).

**Comment by bradhilton** [comment]
@ValeKnappich the logprobs are not used for the kl divergence, but are actually used [here](https://github.com/OpenPipe/ART/blob/f36cbcb97b278dcbda753481f407b8ea081c5f42/src/art/local/train.py#L112C9-L132C10) for PPO-style clipping:
```python
        # take the logprobs from the input data (from the Choice objects) as the old/original policy logprobs
        old_logprobs = shift_tensor(inputs["logprobs"], 0.0)
        advantages = shift_tensor(inputs["advantages"], 0.0)
        assistant_mask = shift_tensor(inputs["assistant_mask"], False).to(
            new_logprobs.dtype
        )
        weights = shift_tensor(inputs["weights"], 0.0)
        # Assume missing old logprobs were sampled under the current policy
        old_logprobs = torch.where(
            torch.isnan(old_logprobs),
            new_logprobs,
            old_logprobs,
        )
        # Calculate the probability ratio between the latest version of the model and old/original version of the model
        prob_ratio = torch.exp(new_logprobs - old_logprobs)
        epsilon = _config.get("epsilon", 0.2)
        epsilon_high = _config.get("epsilon_high", epsilon)
        if epsilon_high is None:
            epsilon_high = epsilon
        # Constrain the probability ratio so that the new version of the model does not stray too far from the original version
        policy_loss = -torch.min(
            prob_ratio * advantages,
            torch.clip(prob_ratio, 1 - epsilon, 1 + epsilon_high) * advantages,
        )
```

**Comment by ValeKnappich** [comment]
@bradhilton I see, that makes more sense.

My original problem is solved, so I will close this issue now. I think it might be helpful to add some more documentation or more descriptive errors to avoid the confusion regarding the expected format in `messages_and_choices`, e.g. in the lines of:

```py
assert not any(msg["role"] == "assistant" msg for msg in trajectory.messages_and_choices if isinstance(msg, dict)), "Found assistant message without Choice wrapper."
```
